lucet-runtime/lucet-runtime-internals/src/error.rs

```diff
@@ -1,81 +1,81 @@
+#[derive(Debug, Error)]
 pub enum Error {
-    #[fail(display = "Invalid argument: {}", _0)]
+    #[error("Invalid argument: {0}")]
     InvalidArgument(&'static str),
 
     /// A [`Region`](trait.Region.html) cannot currently accommodate additional instances.
-    #[fail(display = "Region capacity reached: {} instances", _0)]
+    #[error("Region capacity reached: {0} instances")]
     RegionFull(usize),
 
     /// A module error occurred.
-    #[fail(display = "Module error: {}", _0)]
+    #[error("Module error: {0}")]
     ModuleError(ModuleError),
 
     /// A method call or module specification would exceed an instance's
     /// [`Limit`s](struct.Limits.html).
-    #[fail(display = "Instance limits exceeded: {}", _0)]
+    #[error("Instance limits exceeded: {0}")]
     LimitsExceeded(String),
 
     /// A method call attempted to modify linear memory for an instance that
     /// does not have linear memory
-    #[fail(display = "No linear memory available: {}", _0)]
+    #[error("No linear memory available: {0}")]
     NoLinearMemory(String),
 
     /// An attempt to look up a WebAssembly function by its symbol name failed.
-    #[fail(display = "Symbol not found: {}", _0)]
+    #[error("Symbol not found: {0}")]
     SymbolNotFound(String),
 
     /// An attempt to look up a WebAssembly function by its table index failed.
-    #[fail(display = "Function not found: (table {}, func {})", _0, _1)]
+    #[error("Function not found: (table {0}, func {1}")]
     FuncNotFound(u32, u32),
 
     /// An instance aborted due to a runtime fault.
-    #[fail(display = "Runtime fault: {:?}", _0)]
+    #[error("Runtime fault: {0:?}")]
     RuntimeFault(FaultDetails),
 
     /// An instance terminated, potentially with extra information about the termination.
     ///
     /// This condition can arise from a hostcall explicitly calling
     /// [`Vmctx::terminate()`](vmctx/struct.Vmctx.html#method.terminate), or via a custom signal handler
     /// that returns [`SignalBehavior::Terminate`](enum.SignalBehavior.html#variant.Terminate).
-    #[fail(display = "Runtime terminated")]
+    #[error("Runtime terminated")]
     RuntimeTerminated(TerminationDetails),
 
     /// IO errors arising during dynamic loading with [`DlModule`](struct.DlModule.html).
-    #[fail(display = "Dynamic loading error: {}", _0)]
-    DlError(#[cause] std::io::Error),
+    #[error("Dynamic loading error: {0}")]
+    DlError(#[from] std::io::Error),
 
-    #[fail(display = "Instance not returned")]
+    #[error("Instance not returned")]
     InstanceNotReturned,
 
-    #[fail(display = "Instance not yielded")]
+    #[error("Instance not yielded")]
     InstanceNotYielded,
 
-    #[fail(display = "Start function yielded")]
+    #[error("Start function yielded")]
     StartYielded,
 
+    // A test in instruction_counting.rs requires the conversion of a
+    // Lucetc error to the error type in this crate.
+    #[error("Lucetc error")]
+    LucetcError(#[from] LucetcError),
+
```

Comment:
I'm truly not keen on the hackery that I've done here and would love thoughts on another approach.